File: 02_separate_chaining.py
class Node:
    def __init__(self, key):
        self.key = key
        self.next = None

class SeparateChainingHashTable:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size

    def hash_function(self, key):
        return key % self.size

    def insert(self, key):
        idx = self.hash_function(key)
        node = Node(key)
        node.next = self.table[idx]
        self.table[idx] = node

    def search(self, key):
        idx = self.hash_function(key)
        curr = self.table[idx]
        while curr:
            if curr.key == key:
                return True
            curr = curr.next
        return False

if __name__ == "__main__":
    h = SeparateChainingHashTable(7)
    for k in [10, 20, 5, 15, 7]:
        h.insert(k)
    print("Search 15:", h.search(15))
    print("Search 99:", h.search(99))

File: 02_separate_chaining.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 7

struct Node {
    int key;
    struct Node* next;
};

struct Node* hashTable[SIZE];

int hashFunction(int key) {
    return key % SIZE;
}

void insert(int key) {
    int idx = hashFunction(key);
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->key = key;
    newNode->next = hashTable[idx];
    hashTable[idx] = newNode;
}

int search(int key) {
    int idx = hashFunction(key);
    struct Node* curr = hashTable[idx];
    while (curr) {
        if (curr->key == key) return 1;
        curr = curr->next;
    }
    return 0;
}

int main() {
    int keys[] = {10, 20, 5, 15, 7};
    for (int i = 0; i < 5; i++)
        insert(keys[i]);

    printf("Search 15: %d\n", search(15));
    printf("Search 99: %d\n", search(99));
    return 0;
}
